### ISSUE TITLE
Some refactorings to abstract from C and PHP

### ISSUE THREAD

**Issue by einheitlix** [fabsx00/joern#80]
As discussed in PR #79, I would suggest refactoring some classes within the `ast` package in order to abstract from C and from PHP. Basically every class in `ast` itself should be seen as a base class that can be extended by C- or PHP-aware implementations in `ast.c` resp. `ast.php` packages (the same goes, of course, for packages such as `ast.expressions`, `ast.declarations`, etc.).

My basic approach would be to say, "whenever a base class declares a field that is only used for C (resp. only for PHP), create a new extending class, declare that field in the extending class only, and move associated get- resp. set- methods to the extending class accordingly."

Another problem: Often, the C implementations override the methods `addChild(ASTNode)` and `accept(ASTNodeVisitor)`, neither of which is needed by the PHP classes. I am not sure what to do about these. The absolutely 100% clean way would be to also move them to extending classes, but this would imply having to create C-aware extending implementations for each and every single class in the `ast` package. Therefore, as long as they do not hurt the PHP world, we can perhaps simply leave them in the base classes, and only move them to extending C-aware classes when such an extending class is declared anyway, due to the approach mentioned above.

Note that `accept(ASTNodeVisitor)` never hurts PHP world. At the contrary, `addChild(ASTNode)` can be confusing for PHP world, because in PHP world we *do* need the super-method `ASTNode.addChild(ASTNode)` in order to add a node to another node's list of children, but we do *not* need the `addChild(ASTNode)` override declared in most classes that inherit from `ASTNode` and which call various handlers which are not needed for PHP and can have undesired side effects. The override, however, prevents us from calling the super-method `ASTNode.addChild(ASTNode)` directly in PHP world. Currently this is solved by having the edge interpreter in PHP call the various handler methods directly, and have the handler methods call the super-method `ASTNode.addChild(ASTNode)`.

All in all, attending to this issue will split the C and the PHP world cleanly and more easily allow for integration of further languages into Joern. I will add one comment to this thread per class that I think needs refactoring. Once that refactoring is done, the corresponding comment can be deleted, or marked in some way. :blush:

**Comment by einheitlix** [comment]
`ast.expressions.Identifier`

* Declare `ast.php.expressions.PHPIdentifier`
* Move field `ASTNode name` to `PHPIdentifier`, get- and set- methods accordingly.

**Comment by einheitlix** [comment]
`ast.functionDef.FunctionDef`

* Declare `ast.c.functionDef.CFunctionDef`
    * Move field `Identifier identifier` to `CFunctionDef`, get- and set- methods accordingly (a function's name is a property of the node in PHP world)
    * Move `getEscapedCodeStr()`, `getFunctionSignature()`, `accept(ASTNodeVisitor)`, `addChild(ASTNode)` to `CFunctionDef`.

* Declare `ast.php.functionDef.PHPFunctionDef`
    * Move field `Identifier returnType` to `PHPFunctionDef`, get- and set- methods accordingly (just out of curiosity: how is a function's return type represented in C world?)
    * Move `getName()` and `setName()` to `PHPFunctionDef` (this retrieves the function's name, stored as a property.)
    * Move `getDocComment()` and `setDocComment()` to `PHPFunctionDef` (this retrieves the function's doccomment, stored as a property; unsupported in C world.)

**Comment by einheitlix** [comment]
Side comment: `ast.php.functionDef.{TopLevelFunctionDef,Closure,Method}` should extend `ast.php.functionDef.PHPFunctionDef` instead of `ast.functionDef.FunctionDef` once the previous comment has been tackled.

**Comment by einheitlix** [comment]
`ast.functionDef.Parameter`

* Declare `ast.c.functionDef.CParameter`
    * Move field `ParameterType type` to `CParameter`, get- and set-methods accordingly
    * Move field `Identifier identifier` to `CParameter`, get- and set-methods accordingly
    * Move ` addChild(ASTNode)` and `accept(ASTNodeVisitor)` to `CParameter`

* `ast.php.functionDef.PHPParameter` already exists and declares all fields it needs by itself, no changes necessary.

**Comment by einheitlix** [comment]
`ast.statements.blockstarters.IfStatement`

* Declare `ast.c.statements.blockstarters.CIfStatement`
    * Move field `ElseStatement elseNode` to `CIfStatement`, get- and set-methods accordingly
    * Move overrides `getChildCount()`, `getChild(int)`, `accept(ASTNodeVisitor)` to `CIfStatement`

* Refactorings in the already-existing `ast.php.statements.blockstarters.PHPIfStatement`:
    * Remove re-overrides `getChildCount()`, `getChild(int)`, used to restore  `ASTNode`'s original behavior
    * Remove `getElseNode()` and `setElseNode(ElseStatement)`, these methods can now simply be undeclared instead of throwing a `RuntimeException`

**Comment by einheitlix** [comment]
That's all for now, the other ones should be fine, except for that `addChild(ASTNode)` override in most classes extending `ASTNode`. These overrides are only needed in C world, and may have undesired side effects in PHP world, as mentioned in the initial comment. But perhaps we can take care of problems due to this thing if and when they occur. :smile:

**Comment by einheitlix** [comment]
`ast.statements.jump.ReturnStatement`

* For C world:
    * Overide method `addChild(ASTNode)` and have it call `setReturnExpression(ASTNode)`
    * Maybe refactor C world to call `getReturnExpression()` on return statements rather than `getChild(0)`

Also see PR #81.

**Comment by einheitlix** [comment]
`ast.logical.statements.Label`

* For C world:
    * Overide method `addChild(ASTNode)` and have it call `setNameChild(ASTNode)`
    * Maybe refactor C world to call `getNameChild()` on label statements rather than `getChild(0)`

Also see PR #81.

**Comment by einheitlix** [comment]
`ast.statements.jump.GotoStatement`

* For C world:
    * Overide method `addChild(ASTNode)` and have it call `setTargetLabel(ASTNode)`
    * Maybe refactor C world to call `getTargetLabel()` on goto statements rather than `getChild(0)`
    * Change method `getTargetName()` to return `getTargetLabel().getEscapedCodeStr()` instead of `getChild(0).getEscapedCodeStr()`

Also see PR #81.

**Comment by einheitlix** [comment]
`ast.statements.jump.ThrowStatement`

* For C world:
    * Overide method `addChild(ASTNode)` and have it call `setThrowExpression(ASTNode)`
    * Maybe refactor C world to call `getThrowExpression()` on throw statements rather than `getChild(0)`

Also see PR #82.

**Comment by einheitlix** [comment]
`ast.expressions.ConditionalExpression`

* For C world:
    * Overide method `addChild(ASTNode)` and have it call `setCondition(ASTNode)`, `setTrueExpression(ASTNode)` or `setFalseExpression(ASTNode)` according to what node is currently being added.
    * Maybe refactor C world to call `getCondition()`, `getTrueExpression()` or `getFalseExpression()` on conditional expressions rather than `getChild(0)`, `getChild(1)` or `getChild(2)`, respectively.

Also see PR #82.

**Comment by einheitlix** [comment]
`ast.statements.blockstarters.CatchStatement`

* For C world:
    * Overide method `addChild(ASTNode)` and have it call `setExceptionIdentifier(Identifier)`, `setVariableName(ASTNode)` or `setContent(CompoundStatement)` according to what node is currently being added.
    * Maybe refactor C world to call `getExceptionIdentifier()`, `getVariableName()` or `getContent()` on catch statements rather than `getChild(0)`, `getChild(1)` or `getChild(2)`, respectively.

Also see PR #82.

**Comment by einheitlix** [comment]
`ast.statements.blockstarters.TryStatement`

* For C world:
    * Overide method `addChild(ASTNode)` and have it call `setContent(CompoundStatement)` or `getCatchList().addCatchStatement(CatchStatement)`according to what node is currently being added.
    * Maybe refactor C world to call `getContent()` or `getCatchList().getCatchStatement(i)` on try statements rather than `getChild(0)` or `getChild(i+1)`, respectively.

Also see PR #82.

**Comment by einheitlix** [comment]
`ast.expressions.CallExpression`

* Declare `ast.c.expressions.CCallExpression`
    * Move `addChild(ASTNode)` to `CCallExpression`.

Also see PR #83.

**Comment by einheitlix** [comment]
`ast.expressions.ArrayIndexing`

* For C world:
    * Overide method `addChild(ASTNode)` and have it call `setArrayExpression(ASTNode)` or `setArrayIndex(ASTNode)` according to what node is currently being added.
    * Maybe refactor C world to call `getArrayExpression()` or  `getArrayIndex()` on array indexing expressions rather than `getChild(0)` or `getChild(1)`, respectively.

Also see PR #85.

**Comment by einheitlix** [comment]
`ast.expressions.SizeofExpr`

* Only cosmetic:
    * Rename class `SizeofExpr` to `SizeofExpression` and adjust C world accordingly.

Also see PR #86.

**Comment by einheitlix** [comment]
`ast.expressions.UnaryExpression`
and `ast.expressions.UnaryOperationExpression` which extends it

* For C world:
    * Override method `addChild(ASTNode)` in `ast.expressions.UnaryExpression` and have it call `setExpression(ASTNode)`.
    * Maybe refactor C world to call `getExpression()` rather than `getChild(0)`.

Also see PR #88.

**Comment by einheitlix** [comment]
`ast.expressions.CastExpression`

* For C world:
    * Maybe refactor C world to call `getCallExpression()` rather than `getChild(1)`.

Also see PR #89.

**Comment by einheitlix** [comment]
`ast.expressions.PostIncDecOperationExpression`
(previously called `ast.expressions.IncDecOp`)

* For C world:
    * Override method `addChild(ASTNode)` and have it call `setVariableExpression(ASTNode)`.
    * Maybe refactor C world to call `getVariableExpression()` rather than `getChild(0)`.

Also see PR #89.

**Comment by einheitlix** [comment]
Commits 0f640286d54d1718f680b0c8a9c77a64f7f0bc52, 7b29150c671cb8103ecf56a409083d069a144397, fa0e73dafde21217c1c30c08d24d20958de39d7e, 3a329f80915ce3e3daa151f49d2ee127b8b1bbde, c44dba2da08c370f9a1864510a12b6780f330e1e, and 3c0f6fed956bd00d95c116e42254be2d3c04ddb9 take care of all the problems described here, so I'm closing this issue.

Be aware that refactorization implied renaming some AST nodes. C world now (like PHP world) also has to use some "specific" types instead of the generic ones:
* `FunctionDef` -> `CFunctionDef`
* `Parameter` -> `CParameter`
* `ParameterType` -> `CParameterType`
* `IfStatement` -> `CIfStatement`
* `ElseStatement` -> `CElseStatement`
* `CallExpression` -> `CCallExpression`
* `SizeOfExpr` -> `SizeOfExpression` (ok, this last one is only cosmetic, for naming consistency. :wink:)

It is all taken care of in Joern, but you will have to adapt python-joern & co. accordingly.
